src/cpp/fastdds/subscriber/DataReaderImpl.cpp

```diff
@@ -415,13 +415,19 @@ void DataReaderImpl::InnerDataReaderListener::on_liveliness_changed(
         RTPSReader* /*reader*/,
         const fastrtps::LivelinessChangedStatus& status)
 {
-    if (data_reader_->listener_ != nullptr)
+    data_reader_->update_liveliness_status(status);
+    DataReader* user_reader = data_reader_->user_datareader_;
+
+    if ( (data_reader_->listener_ != nullptr) &&
+            (user_reader->get_status_mask().is_active(StatusMask::liveliness_changed())))
     {
-        data_reader_->listener_->on_liveliness_changed(data_reader_->user_datareader_, status);
+        LivelinessChangedStatus reader_status;
+        data_reader_->get_liveliness_changed_status(reader_status);
+        data_reader_->listener_->on_liveliness_changed(user_reader, reader_status);
     }
     else
     {
-        data_reader_->subscriber_->subscriber_listener_.on_liveliness_changed(data_reader_->user_datareader_, status);
+        data_reader_->subscriber_->subscriber_listener_.on_liveliness_changed(user_reader, status);
```

Comment:
I'd pass the internal updated liveliness_changed_status_ to the parents. I know they are not using it, and reading from the `reader` instead, but somehow it feels wrong sending the status that is not managed at DDS level.